### PR TITLE
feat(transpile,build-plugin) add sourcemaps and watch support for plugin development

### PR DESCRIPTION
**Pull request by MichaelPetrinolis**
take into account --watch when running build-plugin.
include sourcemaps based on environment configuration

### CODE REVIEW

File: skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js [hunk 1]
```diff
@@ -49,13 +62,30 @@ export function buildPluginJavaScript(dest, format) {
...
+
+    let buildOptions = new Configuration(project.build.options, defaultBuildOptions);
+    let sourcemaps = buildOptions.isApplicable('sourcemaps');
+
     return gulp.src(project.plugin.source.js)
+      .pipe(gulpIf(sourcemaps, gulpSourcemaps.init()))
       .pipe(transpile)
+      .pipe(gulpIf(sourcemaps, gulpSourcemaps.write('.', {
+        includeContent: false,
+        sourceRoot: '../../src'
```
Comment (3cp) [inline skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js]: Is `'../../src'` only for the use case of your linked npm package debug?

That sourceRoot seems irrelevant and also harmful to general use case.

File: skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js [hunk 2]
```diff
@@ -49,13 +62,30 @@ export function buildPluginJavaScript(dest, format) {
...
     const transpile = babel(opts);
+
+    let defaultBuildOptions = {
+      minify: 'stage & prod',
+      sourcemaps: 'dev & stage',
+      rev: false
+    };
+
+    let buildOptions = new Configuration(project.build.options, defaultBuildOptions);
+    let sourcemaps = buildOptions.isApplicable('sourcemaps');
```
Comment (3cp) [inline skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js]: This seems unnecessary. Why not just turn on sourcemap without condition? That's what core aurelia modules are trying to do now, shipping sourcemap in all the builds.

It's app bundler's job to decide whether to use the source map in dev build, or strip the source map in production build. All bundlers are capable of doing that.

File: skeleton/plugin/aurelia_project/tasks/build-plugin.ext [hunk 3]
```diff
@@ -10,12 +10,13 @@ import {pluginMarkup} from './process-markup';
 import {pluginCSS} from './process-css';
 import {pluginJson} from './process-json';
 import {buildPluginJavaScript} from './transpile';
+import { CLIOptions } from 'aurelia-cli';
 
 function clean() {
-  return del('dist');
+  return del(['dist/**', '!dist']);
```
Comment (3cp) [inline skeleton/plugin/aurelia_project/tasks/build-plugin.ext]: It seems unnecessary. Is that removing of 'dist/' folder upsetting some of your tools/editors?

File: skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js (hunk 1 above)
Comment (3cp) [reply skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js]: As I said before, `.pipe(gulp.dest('dist', {sourcemaps: '.'})` seems working fine with default included source contents. Some trouble with your debug session on that simpler setup?

Comment (MichaelPetrinolis) [reply skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js]: if you inline the sources, you are no more able to modify the original source files while debugging

File: skeleton/plugin/aurelia_project/tasks/build-plugin.ext (hunk 3 above)
Comment (MichaelPetrinolis) [reply skeleton/plugin/aurelia_project/tasks/build-plugin.ext]: yes, it creates some problems in rebuild. I will not delete dist if the --watch switch is enabled

File: skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js (hunk 1 above)
Comment (3cp) [reply skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js]: Oh I see.

File: skeleton/plugin/aurelia_project/tasks/build-plugin.ext (hunk 3 above)
Comment (3cp) [reply skeleton/plugin/aurelia_project/tasks/build-plugin.ext]: You probably can skip clean step in non-first build in watch mode.

**Comment by 3cp** [comment]
The build-plugin task can be cleaned up to only do "clean" at beginning.

```js
import gulp from 'gulp';
import del from 'del';
import { pluginMarkup } from './process-markup';
import { pluginCSS } from './process-css';
import { pluginJson } from './process-json';
import { buildPluginJavaScript } from './transpile';
import { CLIOptions } from 'aurelia-cli';

function clean() {
  return del('dist');
}

let build = gulp.series(
  gulp.parallel(
    // package.json "module" field pointing to dist/native-modules/index.js
    pluginMarkup('dist/native-modules'),
    pluginCSS('dist/native-modules'),
    pluginJson('dist/native-modules'),
    buildPluginJavaScript('dist/native-modules', 'es2015'),

    // package.json "main" field pointing to dist/native-modules/index.js
    pluginMarkup('dist/commonjs'),
    pluginCSS('dist/commonjs'),
    pluginJson('dist/commonjs'),
    buildPluginJavaScript('dist/commonjs', 'commonjs'),
  ),
  done => {
    console.log('Finish building Aurelia plugin to dist/commonjs and dist/native-modules.');
    done();
  }
);

let main;
if (CLIOptions.hasFlag('watch')) {
  main = gulp.series(
    clean,
    () => {
      console.log('Watching plugin sources for changes ...');
      return gulp.watch('src/**/*', { ignoreInitial: false }, build);
    }
  );
} else {
  main = gulp.series(
    clean,
    build
  );
}

export { main as default };
```

File: skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js (hunk 1 above)
Comment (3cp) [reply skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js]: Also forgot.

Please add a tailing `/` to `sourceRoot: '../../src/'`. Based on sourcemap spec, "This value is prepended to the individual entries in the “source” field.".

I had encountered some libs before that strictly follow the spec, it leads to `../../srcsource/file/path` without probably inserting `/`.

The tailing `/` can avoid those subtle bugs.

Comment (3cp) [reply skeleton/cli-bundler/aurelia_project/tasks__if_babel/transpile.js]: There is another small problem.

The package.json has
```
"files": [
    "dist"
  ],
```

It means by default, it only publish dist folder to npm package, not the src folder. We need to update package.json if we don't have embedded source in sourcemaps.

```
"files": [
    "dist",
    "src"
  ],
```

**Comment by MichaelPetrinolis** [comment]
A concern of mine is that if you npm install a plugin as a file, and you already watch the application, then it might have some issues deleting the dist folder (start watching the plugin after watching the app).

**Comment by 3cp** [comment]
`au run` watch generates files in scripts/ folder. The dev-app running mode does not use dist/ files at all.

`au build-plugin --watch` generates files in dist/ folder.

They should be fine running together.

**Comment by MichaelPetrinolis** [comment]
I mean in another Aurelia app, where you have installed the Plugin, like the setup mentioned in https://github.com/aurelia/cli/pull/1124#issuecomment-506683999

**Comment by 3cp** [comment]
I did some test on gulp sourcemap, it looks like your sourceRoot is much needed for correct source path no matter includeContent or not.

The default gulp v4 offering is incorrect. https://github.com/gulpjs/gulp/issues/2288#issuecomment-506953894

**Comment by 3cp** [comment]
LOL, I am looking back to one of my private aurelia plugin to see why my plugin is not affected by this gulp source path issue. I actually did a hack to remove duplicated paths, which I totally forgot:-)

```js
.pipe(through2.obj(function(file, enc, cb) {
    if (file.isBuffer() && file.sourceMap) {
      file.sourceMap.file = file.basename;
      file.sourceMap.sources = file.sourceMap.sources.map(n => path.basename(n));
    }
    cb(null, file);
  }))
  .pipe(gulp.dest('dist', {sourcemaps: '.'}));
```

So `element/hello-world.js` will have `hello-world.js` in both `sources` and `file` field, this avoid duplicated `element/element/` path. But my source path is still not fully correct (because it starts with dist/ not src/), but that's not an issue for me as I don't run debugger.

**Comment by 3cp** [comment]
This looks good to merge.
@MichaelPetrinolis has you tested your concern over the dist deleting?

**Comment by MichaelPetrinolis** [comment]
@3cp It seems to be ok.
